Fix rails restart issue with Puma

prathamesh-sonpatki opened the pull request:
- We need to pass the restart command to Puma so that it will use it
  while restarting the server.
- Besides that we need to remove the pid file for the previous running
  server because otherwise Rack complains about it's presence.
- This also requires some changes on Puma side which are being tracked
  here - https://github.com/puma/puma/pull/936.
- Fixes #23910.

r? @rafaelfranca cc @evanphx

rubys commented on railties/lib/rails/commands/server.rb:
```diff
@@ -94,7 +94,8 @@ def default_options
         caching:            nil,
-        pid:                Options::DEFAULT_PID_PATH
+        pid:                Options::DEFAULT_PID_PATH,
+        restart_cmd:        'bin/rails s'
```
Question: if I start the rails server using, say:

    rails server -b 0.0.0.0 -p 4567

And then restart the server using:

    rails touch

Will the same options be used?

prathamesh-sonpatki replied:
@rubys No. It will start on port 3000 instead of 4567.

```
$ rails server -b 0.0.0.0 -p 4567
=> Booting Puma
=> Rails 5.0.0.beta3 application starting in development on http://0.0.0.0:4567
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.2.0 (ruby 2.3.0-p0), codename: Spring Is A Heliocentric Viewpoint
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:4567
Use Ctrl-C to stop
* Restarting...
=> Booting Puma
=> Rails 5.0.0.beta3 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.2.0 (ruby 2.3.0-p0), codename: Spring Is A Heliocentric Viewpoint
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://localhost:3000
* Closing unused inherited connection: tcp://0.0.0.0:4567
Use Ctrl-C to stop
```

I hadn't thought of this scenario. I will check how to make it work.

prathamesh-sonpatki replied:
@rubys I fixed it here - https://github.com/rails/rails/pull/24331/commits/233798977f5b01016e7ce40aa50c547a183bf9b8.

Now it works as expected.

```
$ rails server -b 0.0.0.0 -p 4567
=> Booting Puma
=> Rails 5.0.0.beta3 application starting in development on http://0.0.0.0:4567
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.2.0 (ruby 2.3.0-p0), codename: Spring Is A Heliocentric Viewpoint
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:4567
Use Ctrl-C to stop
* Restarting...
=> Booting Puma
=> Rails 5.0.0.beta3 application starting in development on http://0.0.0.0:4567
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.2.0 (ruby 2.3.0-p0), codename: Spring Is A Heliocentric Viewpoint
* Min threads: 5, max threads: 5
* Environment: development
* Inherited tcp://0.0.0.0:4567
Use Ctrl-C to stop
```

kaspth commented on railties/lib/rails/dev_caching.rb:
```diff
@@ -15,6 +15,7 @@ def enable_by_file
         end
 
         FileUtils.touch 'tmp/restart.txt'
+        FileUtils.rm_f('tmp/pids/server.pid')
```
Let's extend the existing test which asserts that the temp file has been removed to also assert that the pid file is removed.

prathamesh-sonpatki replied:
Added coverage.

kaspth commented on railties/lib/rails/tasks/restart.rake:
```diff
@@ -2,4 +2,5 @@ desc "Restart app by touching tmp/restart.txt"
 task :restart do
   FileUtils.mkdir_p('tmp')
   FileUtils.touch('tmp/restart.txt')
+  FileUtils.rm_f('tmp/pids/server.pid')
```
Ditto here.

prathamesh-sonpatki replied:
Added coverage.

kaspth commented on railties/test/commands/server_test.rb:
```diff
@@ -118,4 +118,17 @@ def test_default_options
+
+  def test_restart_command_contains_customized_options
+    args = ["-p", "4567"]
+    ARGV.replace args
```
Is this inline mutation a good idea? Wouldn't at least be better to capture the old ARGV value and reset in ensure.

rafaelfranca replied:
:+1:

kaspth commented on railties/lib/rails/commands/server.rb:
```diff
@@ -130,5 +131,9 @@ def log_to_stdout
       end
+
+      def restart_command
+        "bin/rails s #{ARGV.join(' ')}"
```
Can we just expand the `s` to `server`? 😁

rafaelfranca replied:
:+1:

prathamesh-sonpatki replied:
Done.

kaspth commented on railties/lib/rails/commands/server.rb:
```diff
@@ -94,7 +94,8 @@ def default_options
         caching:            nil,
-        pid:                Options::DEFAULT_PID_PATH
+        pid:                Options::DEFAULT_PID_PATH,
+        restart_cmd:        restart_command
```
Is `restart_cmd` guaranteed to work for other servers people might use in dev?

kaspth replied:
By "work" I mean that they won't blow up on unsupported keys for instance (not that they will automatically use the generated command).

prathamesh-sonpatki replied:
I tested on Thin, Unicorn and Webrick. The `restart_cmd` has no effect on them. Meaning `rails restart` continues to **not work** while using those servers as before.

prathamesh-sonpatki replied:
Also these servers don't blow up on unsupported key `restart_cmd`.

prathamesh-sonpatki committed:
Fix rails restart issue with Puma

- We need to pass the restart command to Puma so that it will use it
  while restarting the server.
- Also made sure that all the options passed by user while starting
  the server are used in the generated restart command so that they will
  be used while restarting the server.
- Besides that we need to remove the server.pid file for the previous running
  server because otherwise Rack complains about it's presence.
- We don't care if the server.pid file does not exist. We only want to delete
  it if it exists.
- This also requires some changes on Puma side which are being tracked
  here - https://github.com/puma/puma/pull/936.
- Fixes #23910.

prathamesh-sonpatki commented:
@kaspth Thanks for the review! I updated the code.

prathamesh-sonpatki commented on railties/test/commands/server_test.rb:
```diff
@@ -118,4 +118,18 @@ def test_default_options
+
+    assert_equal expected, server.default_options[:restart_cmd]
+  ensure
+    ARGV.replace original_args
```
Made sure that original ARGV are replaced at the end of test. The other option was to use test specific constant TEST_ARGV like SPEC_ARGV in Puma tests but I preferred this option over that.

prathamesh-sonpatki replied:
The other option is to use `ARGV[0..-1] = args` instead of `ARGV.replace`. 

@rafaelfranca What do you think about `ARGV[0..-1] = args`?

rafaelfranca replied:
`replace` is fine to me.

rafaelfranca commented:
Tests are broken.

prathamesh-sonpatki commented:
Thanks, I fixed the tests.

kaspth commented:
❤️🤘

steakknife commented:
This should be backported to stable branches. It's a minimal change needed ASAP.

https://github.com/puma/puma/issues/1117

rafaelfranca commented:
Which stable branch are you talking about? This is already in 5.0.0.

prathamesh-sonpatki commented:
I looked into this issue. With Puma 3.5.0 and above, I am getting this error, https://gist.github.com/prathamesh-sonpatki/bc0cac8929dbf17316e8b4b7dda93e20.

With Puma 3.4 and 3.3, the `rails restart` command is working as expected. I will look into why it stopped working with Puma 3.5.

prathamesh-sonpatki commented:
I have opened https://github.com/puma/puma/pull/1120 to fix this issue for Pume 3.5 and above.
